### ISSUE TITLE
The threaded-video defect is not gl3-only: 22 call sites across every driver, and one of them is a free

### ISSUE THREAD

**Issue by maxengel** [maxengel/rocknix#225]
**Maintainer, 2026-09-19 (chat), after the root cause in #211:**

> Is this something that we could perhaps encounter elsewhere in the codebase? If it's a similar bug to what we've had previously, it would be good to investigate whether there may be other places where this bug or a related bug could be hiding.

Yes, and wider than the crash suggests.

## The class

**An object with automatic storage whose address is handed to another thread, where the dispatching side decides whether to run the work using state the *posting* side is responsible for clearing.** #211's crash is one instance; the mechanism is in `video_thread_loop`, not in gl3.

## Every driver is exposed, not just glcore

`video_thread_texture_handle` is called from **22 sites**, each passing the address of a local:

| File | Sites |
| --- | --- |
| `gfx/drivers/gl3.c` | 2 (the one that crashed) |
| `gfx/drivers/gl2.c`, `gl1.c` | 4 |
| `gfx/drivers/vulkan.c` | 1 |
| `gfx/drivers/d3d8.c`, `d3d9cg.c`, `d3d9hlsl.c`, `d3d10.c`, `d3d11.c`, `d3d12.c` | 12 |
| `gfx/drivers/rsx_gfx.c` | 2 |
| **`gfx/font_driver.c`** | **1 — see below** |

Our handhelds run `glcore`, so gl3 is what we hit. Anyone on gl2, vulkan or a Direct3D backend is exposed to the identical defect, because the fault is in the shared dequeue.

## The font path is worse than the texture path

`font_driver_release_renderer_state` builds `font_free_cmd_t cmd;` on its stack and posts `font_driver_free_wrap`, which calls:

```c
cmd->renderer->free(cmd->renderer_data, cmd->is_threaded);
```

A doubled dispatch there is a **double free**, not a bad read. That does not segfault where it happens — it corrupts the heap and crashes later, somewhere unrelated, with a backtrace that names an innocent frame. If any of #79's unexplained crashes were this, they would look exactly as they did: sudden, unreproducible, and pointing nowhere.

## Why the fix covers all of it

The one-line gate proposed in #211 lives in `video_thread_loop`, above every call site:

```c
have_cmd = (thr->send_cmd != CMD_VIDEO_NONE);
...
if (have_cmd && video_thread_handle_packet(thr, &pkt))
```

So fixing the dequeue fixes all 22, including the free.

## The same class in our own code — leads, not findings

EmulationStation runs threads too, and the equivalent shape is a **detached** thread holding a pointer to something the interface may destroy. We have already had one (the cloud wizard page: a callback retained a page after `GuiSettings::close()` had deleted it, 2026-09-05).

Present: `std::thread`/`pthread_create` in `ThreadedCloudSync`, `ThreadedHasher`, `ThreadedBluetooth`, `ThreadedScraper`, `ContentInstaller`, `GuiAutoScrape`, `GuiInstall`, `GuiNetPlay`, `GuiBluetoothPair`, `HttpServerThread`, `VolumeControl`; `.detach()` at `Window.cpp` ×4, `FileSystemUtil.cpp`, `ZaparooSupport.cpp`.

**These are unaudited.** Naming them is not a claim that any is wrong — it is the list to walk. The question for each: *can the object whose address this thread holds be destroyed before the thread touches it, and what does the thread do if it is?*

## Acceptance criteria

- [ ] The `video_thread_loop` gate lands as a fork patch and is reported upstream, with #211's dump as evidence.
- [ ] The report names the font double-free explicitly, since it is the more dangerous instance and is not what anyone reproduced.
- [ ] A regression check exists that would have caught it: a fixture that posts a command and forces a frame update into the reply window, asserting the method runs exactly once.
- [ ] Each detached thread above is walked and recorded as safe-with-reason or fixed; the list is in this issue, ticked one at a time.
- [ ] Anything found in ES gets the same treatment as #211: a dump or an equivalent artifact, not an argument.

Related: #211 (the crash, the dump and the mechanism), #79 (the unexplained crashes, of which one or more may be the font path), libretro/RetroArch#19517 / #19518 (the first bug in the same wrapper).

Co-Authored-By: Claude Opus 5 (1M context) <noreply@anthropic.com>

**Comment by maxengel** [comment]
**Scope cut, with the measurement behind it.**

The maintainer asked to limit this to where RetroArch could be running concurrently. Measured on a guest with a game running:

| | Interface idle | Game running |
| --- | --- | --- |
| EmulationStation alive | yes | **yes**, state `S` |
| ES threads | 15 + 3 audio | **15**, audio released |

So ES does run alongside RetroArch — and it is a **separate process**, so a use-after-free in an ES thread crashes ES, not RetroArch, and cannot produce #211's class of fault. The nineteen ES sites listed above are a different blast radius wearing the same name.

**The ES ASan walk and the TSan pass are therefore dropped from the pre-build list.** They remain worth doing as EmulationStation-stability work; they cannot find another #211. What ES and RetroArch actually share during a game is files (saves, already governed by the two-writers rules) and the proxy — not memory, and not the GPU, since ES is asleep and has released audio.

The intersection that mattered was inside RetroArch's own process, and patch 0015's gate sits above all 22 call sites, so it is covered.

One supporting observation for the fix's shape: the wrapper already uses consumer-owned flags everywhere else — `frame.updated` is set by the poster and cleared by the video thread, and `texture.frame_updated` likewise. Only the command path inverted it. Upstream's master arrived at the same conclusion (libretro/RetroArch#19577).

**Comment by maxengel** [comment]
**Closed as delivered (2026-09-23).** The scope cut was your call on 2026-09-19 (*"limit this to where RetroArch could be running concurrently"*), measured and recorded in the comment above; it is now register row **D-QA-039**, which it should have been that day. The one acceptance box is met: patch 0015 (`video-thread-wrapper-run-a-command-once`) ships in every build since, its gate sits above all 22 call sites including the font free, and upstream closed libretro/RetroArch#19577 with the same gate.

Your memory today is right on the shape: the *trigger* is the badge image after an award, which is why it surfaced through RetroAchievements; the *defect* was in the shared wrapper loop, which is why one gate covers every driver. Nothing here waited on a decision; it was listed as one on #236 by mistake (blindspot 52).
